lib/yubihsm.c

```diff
@@ -933,6 +925,282 @@ yh_rc yh_finish_create_session_ext(
+           identifier[0], identifier[1], identifier[2], identifier[3],
+           identifier[4], identifier[5], identifier[6], identifier[7]);
+
+  Msg msg;
+  Msg response_msg;
+  uint16_t authkey_id_n = htons(authkey_id);
+
+  // Send CREATE SESSION command
+  msg.st.cmd = YHC_CREATE_SESSION;
+  msg.st.len = sizeof(authkey_id_n) + sizeof(epk_oce);
+
+  memcpy(msg.st.data, &authkey_id_n, sizeof(authkey_id_n));
+  memcpy(msg.st.data + sizeof(authkey_id_n), epk_oce, sizeof(epk_oce));
+
+  rc = send_msg(connector, &msg, &response_msg, new_session->s.identifier);
+  if (rc != YHR_SUCCESS) {
+    goto err;
+  }
+
+  // Parse response
+  if (response_msg.st.cmd != YHC_CREATE_SESSION_R) {
+    rc = translate_device_error(response_msg.st.data[0]);
+    DBG_ERR("Device error %s (%d)", yh_strerror(rc), response_msg.st.data[0]);
+    goto err;
+  }
+
+  DBG_INT(response_msg.st.data, 1, "SessionId: ");
+  DBG_INT(response_msg.st.data + 1, sizeof(epk_oce), "EPK-SD: ");
+  DBG_INT(response_msg.st.data + 1 + sizeof(epk_oce), SCP_PRF_LEN, "Receipt: ");
+
+  uint8_t shsee[32];
+  if (!ecdh_calculate_secret(curve, esk_oce, sizeof(esk_oce),
+                             response_msg.st.data + 1, sizeof(epk_oce), shsee,
+                             sizeof(shsee))) {
+    DBG_ERR("ecdh_calculate_secret(shsee) %s",
+            yh_strerror(YHR_INVALID_PARAMETERS));
+    return YHR_INVALID_PARAMETERS;
+  }
+
+  uint8_t shsss[32];
+  if (!ecdh_calculate_secret(curve, privkey, privkey_len, device_pubkey,
+                             device_pubkey_len, shsss, sizeof(shsss))) {
+    DBG_ERR("ecdh_calculate_secret(shsss) %s",
+            yh_strerror(YHR_INVALID_PARAMETERS));
+    return YHR_INVALID_PARAMETERS;
+  }
+
+  uint8_t shs[4 * SCP_KEY_LEN];
+  x9_63_sha256_kdf(shsee, sizeof(shsee), shsss, sizeof(shsss), sharedInfo,
+                   sizeof(sharedInfo), shs, sizeof(shs));
+
+  DBG_INT(shs, sizeof(shs), "SHS: ");
+
+  uint8_t keys[2 * sizeof(epk_oce)], mac[SCP_PRF_LEN];
+  memcpy(keys, response_msg.st.data + 1, sizeof(epk_oce));
+  memcpy(keys + sizeof(epk_oce), epk_oce, sizeof(epk_oce));
+  compute_full_mac(keys, sizeof(keys), shs, SCP_KEY_LEN, mac);
+
+  if (memcmp(mac, response_msg.st.data + 1 + sizeof(epk_oce), SCP_PRF_LEN)) {
+    DBG_ERR("Verify receipt %s",
+            yh_strerror(YHR_SESSION_AUTHENTICATION_FAILED));
+    rc = YHR_SESSION_AUTHENTICATION_FAILED;
+    goto err;
+  }
+
+  memcpy(new_session->s.s_enc, shs + SCP_KEY_LEN, SCP_KEY_LEN);
+  memcpy(new_session->s.s_mac, shs + 2 * SCP_KEY_LEN, SCP_KEY_LEN);
+  memcpy(new_session->s.s_rmac, shs + 3 * SCP_KEY_LEN, SCP_KEY_LEN);
+  memcpy(new_session->s.mac_chaining_value, mac, SCP_PRF_LEN);
+
+  memset(new_session->s.ctr, 0, SCP_PRF_LEN);
+  increment_ctr(new_session->s.ctr, SCP_PRF_LEN);
+
+  new_session->parent = connector;
+  new_session->authkey_id = authkey_id;
+  new_session->recreate = false;
+  new_session->s.sid = response_msg.st.data[0];
```

Comment:
Hmm can't see that. Maybe another branch ?

Reply:
Found it. But it's missing from yh_begin_create_session_ext as well. Is it a problem if this is wrong ? It will just fail later unless i'm mistaken. Maybe should be removed instead ?